cmd/src/debug_comp.go

```diff
@@ -0,0 +1,228 @@
+	"path/filepath"
+	"strings"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+
+	"golang.org/x/sync/errgroup"
+
+	"golang.org/x/sync/semaphore"
+)
+
+func init() {
+	usage := `
+'src debug compose' invokes docker cli diagnostic commands targeting a containers that are members of a docker-compose network, 
+writing an archive file from their returns. 
+
+Usage:
+
+	src debug compose [command options]
+
+Flags:
+
+	-o					Specify the name of the output zip archive.
+	--no-configs		Don't include Sourcegraph configuration json.
+
+Examples:
+
+	$ src debug compose -o debug.zip
+
+	$ src -v debug compose -no-configs -o foo.zip
+
+`
+
+	flagSet := flag.NewFlagSet("compose", flag.ExitOnError)
+	var base string
+	var noConfigs bool
+	flagSet.StringVar(&base, "o", "debug.zip", "The name of the output zip archive")
+	flagSet.BoolVar(&noConfigs, "no-configs", false, "If true include Sourcegraph configuration files. Default value true.")
+
+	handler := func(args []string) error {
+		if err := flagSet.Parse(args); err != nil {
+			return err
+		}
+
+		// process -o flag to get zipfile and base directory names
+		if base == "" {
+			return fmt.Errorf("empty -o flag")
+		}
+		// declare basedir for archive file structure
+		base, baseDir := processBaseDir(base)
+
+		// init context
+		ctx := context.Background()
+		// open pipe to output file
+		out, err := os.OpenFile(base, os.O_CREATE|os.O_RDWR|os.O_EXCL, 0666)
+		if err != nil {
+			return errors.Wrap(err, "failed to open file")
+		}
+		defer out.Close()
+		// init zip writer
+		zw := zip.NewWriter(out)
+		defer zw.Close()
+
+		//Gather data for safety check
+		containers, err := getContainers(ctx)
+		if err != nil {
+			return errors.Wrap(err, "failed to get containers for subcommand with err")
+		}
+		// Safety check user knows what they are targeting with this debug command
+		log.Printf("This command will archive docker-cli data for %d containers\n SRC_ENDPOINT: %v\n Output filename: %v", len(containers), cfg.Endpoint, base)
+		if verified, _ := verify("Do you want to start writing to an archive?"); !verified {
+			return nil
+		}
+
+		err = archiveCompose(ctx, zw, *verbose, noConfigs, baseDir)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	debugCommands = append(debugCommands, &command{
+		flagSet: flagSet,
+		handler: handler,
+		usageFunc: func() {
+			fmt.Println(usage)
+		},
+	})
+}
+
+// writes archive of common docker cli commands
+func archiveCompose(ctx context.Context, zw *zip.Writer, verbose, noConfigs bool, baseDir string) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	containers, err := getContainers(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to get docker containers")
+	}
+
+	if verbose {
+		log.Printf("getting docker data for %d containers...\n", len(containers))
+	}
+
+	// setup channel for slice of archive function outputs
+	ch := make(chan *archiveFile)
+	g, ctx := errgroup.WithContext(ctx)
+	semaphore := semaphore.NewWeighted(8)
```

Comment:
Interesting use of semaphore! :flags: 

Instead of hardcoding 8, one thing you could consider here would be using Go's maximum "processes", which is usually the CPU count. That would look something like this:

```suggestion
	semaphore := semaphore.NewWeighted(runtime.GOMAXPROCS(0))
```